src/clj/puppetlabs/puppetserver/certificate_authority.clj

```diff
@@ -1314,18 +1314,23 @@
 (schema/defn ensure-subject-alt-names-allowed!
   "Throws an exception if the CSR contains subject-alt-names AND the user has
    chosen to disallow subject-alt-names. Subject alt names can be allowed by
-   setting allow-subject-alt-names to true in the ca.conf file."
+   setting allow-subject-alt-names to true in the ca.conf file. Always allows
+   a single subject alt name that matches the CSR subject, which may be
+   present to comply with RFC 2818 (see SERVER-2338)."
   [csr :- CertificateRequest
    allow-subject-alt-names :- schema/Bool]
   (when-let [subject-alt-names (not-empty (subject-alt-names csr))]
     (if (false? allow-subject-alt-names)
       (let [subject (get-csr-subject csr)]
-        (sling/throw+
-          {:kind :disallowed-extension
-           :msg (format "%s %s %s"
-                        (i18n/tru "CSR ''{0}'' contains subject alternative names ({1}), which are disallowed." subject (str/join ", " subject-alt-names))
-                        (i18n/tru "To allow subject alternative names, set allow-subject-alt-names to true in your ca.conf file.")
-                        (i18n/tru "Then restart the puppetserver and try signing this certificate again."))})))))
+        (if (and (= 1 (count subject-alt-names))
+                 (= (first subject-alt-names) (str "DNS:" subject)))
+          (log/debug "Allowing subject alt name that matches CSR subject.")
+          (sling/throw+
+           {:kind :disallowed-extension
+            :msg (format "%s %s %s"
+                         (i18n/tru "CSR ''{0}'' contains subject alternative names ({1}), which are disallowed." subject (str/join ", " subject-alt-names))
```

Comment:
We probably don't want to include the certname SAN in this error message.

Reply:
why?

Reply:
As a user reading this message, I would interpret this to mean that all the SANs listed were disallowed.

Reply:
I see what you're saying, but I'm not sure how I feel about it. We should maybe discuss?

Reply:
Ah. So, either we list all the SANs for completeness, but make a comment about all but the certname one are allowed, or we filter the certname one from the list of prohibited SANs.

Reply:
Yeah. Not sure which makes more sense. The error is already pretty wordy...

Reply:
Updated with the filtering and a slight rewording, lemme know what you think 🙂